consensus/abci_consensus_connection.go

```diff
@@ -0,0 +1,220 @@
+package consensus
+
+import (
+	"fmt"
+	"bytes"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/lightstreams-network/lightchain/database"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/lightstreams-network/lightchain/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	tmtLog "github.com/tendermint/tendermint/libs/log"
+	ethTypes "github.com/ethereum/go-ethereum/core/types"
+	tmtAbciTypes "github.com/tendermint/tendermint/abci/types"
+	abciTypes "github.com/lightstreams-network/lightchain/consensus/types"
+)
+
+type TendermintABCI struct {
+	db           *database.Database
+	checkTxState *state.StateDB
+	ethRPCClient *rpc.Client
+	logger       tmtLog.Logger
+
+	getCurrentDBState func() (*state.StateDB, error)
+	getCurrentBlock   func() *ethTypes.Block
+}
+
+var _ tmtAbciTypes.Application = &TendermintABCI{}
+
+// NewTendermintABCI creates a new instance of ABCI with clean block's state.
+//
+// ABCI maintains multiple connections (consensus, info, mempool).
+func NewTendermintABCI(db *database.Database, ethRPCClient *rpc.Client) (*TendermintABCI, error) {
+	txState, err := db.Ethereum().BlockChain().State()
+	if err != nil {
+		return nil, err
+	}
+
+	abci := &TendermintABCI{
+		db:                db,
+		ethRPCClient:      ethRPCClient,
+		getCurrentDBState: db.Ethereum().BlockChain().State,
+		getCurrentBlock:   db.Ethereum().BlockChain().CurrentBlock,
+		checkTxState:      txState.Copy(),
+		logger:            log.NewLogger().With("engine", "consensus", "module", "ABCI"),
+	}
+
+	err = abci.ResetBlockState()
+	if err != nil {
+		return nil, err
+	}
+
+	return abci, nil
+}
+
+// InitChain is called upon genesis.
+//
+// Can be used to define validators set and consensus params on the application side.
+//
+// Response:
+//     - If `ResponseInitChain.Validators` is empty, the initial validator set will be the `RequestInitChain.Validators`
+//
+func (abci *TendermintABCI) InitChain(req tmtAbciTypes.RequestInitChain) tmtAbciTypes.ResponseInitChain {
+	abci.logger.Debug("Initializing chain", "chain_id", req.ChainId)
+
+	return tmtAbciTypes.ResponseInitChain{}
+}
+
+// BeginBlock signals the beginning of a new block.
+//
+// Flow:
+//		1. BeginBlock <- ******
+//		2. CheckTx
+//	    3. ExecuteTx
+//		4. EndBlock
+//		5. Commit
+//		6. CheckTx (clean mempool from TXs not included in committed block)
+//
+// The header contains the height, timestamp, and more - it exactly matches the Tendermint block header.
+// The LastCommitInfo and ByzantineValidators can be used to determine rewards and punishments for the validators.
```

Comment:
[MINOR] I feel this comment a bit confusing, maybe a bit better 
```
The header fields `LastCommitInfo` and `Byt.....` 
```

Reply:
except they are not header fields. They are RequestBeginBlock attributes. I will put "req.LastCommitInfo" though to make it clear.